uefi.go

```diff
@@ -45,52 +62,136 @@ func (u UEFIFirmwareDevice) QemuParams(config *Config) []string {
+	secBoot := UEFIFirmwareDevice{
+		Code: filepath.Join(pathBase, pfx+VMFCode+VMFSecboot+VMFSuffix), // OVMF_CODE.secboot.fd
+		Vars: filepath.Join(pathBase, pfx+VMFVars+VMFSecboot+VMFSuffix), // OVMF_VARS.secboot.fd
+	}
+	// SecureBoot+MSVars (amd64 or arm64)
+	secBootVarsMs := UEFIFirmwareDevice{
+		Code: filepath.Join(pathBase, pfx+VMFCode+VMFMs+VMFSuffix), // {O,AA}VMF_CODE.ms.fd
+		Vars: filepath.Join(pathBase, pfx+VMFVars+VMFMs+VMFSuffix), // {O,AA}VMF_VARS.ms.fd
+	}
+	// Insecure+4M
+	insecureBoot4M := UEFIFirmwareDevice{
+		Code: filepath.Join(pathBase, pfx+VMFCode+VMF4MB+VMFSuffix), // OVMF_CODE_4M.fd
+		Vars: filepath.Join(pathBase, pfx+VMFVars+VMF4MB+VMFSuffix), // OVMF_Vars_4M.fd
+	}
+
+	// Insecure
+	insecure := UEFIFirmwareDevice{
+		Code: filepath.Join(pathBase, pfx+VMFCode+VMFSuffix), // OVMF_CODE.fd
+		Vars: filepath.Join(pathBase, pfx+VMFVars+VMFSuffix), // OVMF_Vars.fd
+	}
+
+	var found bool
+	var err error
+	if useSecureBoot {
+		// 4M  and .secboot variants are only on x86
+		switch runtime.GOARCH {
+		case "amd64", "x86_64":
+			found, err = secBoot4MVarsMs.Exists()
+			if err != nil {
+				return &UEFIFirmwareDevice{}, fmt.Errorf("SecureBoot 4M MS Vars erorr: %s", err)
+			}
+			if found {
+				return &secBoot4MVarsMs, nil
+			}
+			found, err = secBoot4M.Exists()
+			if err != nil {
+				return &UEFIFirmwareDevice{}, fmt.Errorf("SecureBoot 4M Vars error: %s", err)
+			}
+			if found {
+				return &secBoot4M, nil
+			}
+			found, err = secBoot.Exists()
+			if found {
+				return &secBoot, nil
+			}
+			if err != nil {
+				return &UEFIFirmwareDevice{}, fmt.Errorf("SecureBoot Vars error: %s", err)
+			}
+
 		}
-		uefiDev.Code = secCode
-		secVars := selectPath(SecVarPaths)
-		if secVars == "" {
-			return &uefiDev, fmt.Errorf("Secureboot requested, but no secureboot Vars file found")
+		found, err = secBootVarsMs.Exists()
+		if err != nil {
+			return &UEFIFirmwareDevice{}, fmt.Errorf("SecureBoot MS Vars error: %s", err)
 		}
-		uefiDev.Vars = secVars
-	} else {
-		codePath := selectPath(UnSecCodePaths)
-		if codePath == "" {
-			codePath = selectPath(SecCodePaths)
+		if found {
+			return &secBootVarsMs, nil
 		}
-		if codePath == "" {
-			return &uefiDev, fmt.Errorf("Failed to find UEFI code firmware")
+
+		return &UEFIFirmwareDevice{}, fmt.Errorf("OVMF secureboot code,vars missing, check: %s", pathBase)
+	}
+
+	switch runtime.GOARCH {
+	case "amd64", "x86_64":
+		found, err = insecureBoot4M.Exists()
+		if err != nil {
+			return &UEFIFirmwareDevice{}, err
 		}
-		uefiDev.Code = codePath
-		unsecVars := selectPath(UnSecVarPaths)
-		if unsecVars == "" {
-			return &uefiDev, fmt.Errorf("Secureboot requested, but no secureboot Vars file found")
+		if found {
+			return &insecureBoot4M, nil
 		}
-		uefiDev.Vars = unsecVars
 	}
-	return &uefiDev, nil
+	found, err = insecure.Exists()
+	if found {
+		return &insecure, nil
+	}
+
+	return &UEFIFirmwareDevice{}, fmt.Errorf("OVMF code,vars missing, check: %s", pathBase)
```

Comment:
Should this be `fmt.Errorf("%s code,vars missing, check: %s", VMFPrefix()+"VMF", pathBase)`

Reply:
Good catch -- and the one further up.  I'll fix.